lib/server/roster.ts

```diff
@@ -0,0 +1,221 @@
+import "server-only";
+
+import { auth } from "@clerk/nextjs/server";
+import { eq } from "drizzle-orm";
+
+import {
+  roster,
+  rosterArtifacts,
+  rosterHeroes,
+  rosterLevels,
+} from "@/drizzle/schema";
+import { drizzleClient } from "@/lib/server/drizzle";
+import { AscensionLevel } from "../characters";
+
+const drizzle = drizzleClient;
+
+export async function createOrUpdateRoster(
+  userId: string,
+): Promise<string | false> {
+  const values = {
+    lastUpdate: new Date().toISOString(),
+    userId: userId,
+  };
+  const createOrUpdateResponse = await drizzle
+    .insert(roster)
+    .values(values)
+    .onConflictDoUpdate({
+      target: [roster.userId],
+      set: { lastUpdate: values.lastUpdate },
+    })
+    .returning({ id: roster.id });
+  return createOrUpdateResponse
+    ? createOrUpdateResponse[0].id.toString()
+    : false;
+}
+
+export type ArtifactData = {
+  artifactId: number;
+  level: number;
+};
+
+export async function createOrUpdateArtifact(
+  rosterId: number,
+  artifact: ArtifactData,
+): Promise<string | false> {
+  const values = {
+    rosterId: rosterId,
+    artifactId: artifact.artifactId,
+    level: artifact.level,
+  };
+  const createOrUpdateResponse = await drizzle
+    .insert(rosterArtifacts)
+    .values(values)
+    .onConflictDoUpdate({
+      target: [rosterArtifacts.rosterId, rosterArtifacts.artifactId],
+      set: { level: values.level },
+    })
+    .returning({ id: rosterArtifacts.id });
+
+  return createOrUpdateResponse
+    ? createOrUpdateResponse[0].id.toString()
+    : false;
+}
+
+export async function createOrUpdateArtifacts(
+  artifacts: ArtifactData[],
+): Promise<string | false> {
+  const { userId } = auth();
+
+  let rosterId: any;
+  if (userId) {
+    rosterId = await createOrUpdateRoster(userId);
+    artifacts.forEach((artifact: any) => {
+      createOrUpdateArtifact(rosterId, artifact);
+    });
```

Comment:
since you are returning the rosterId after this block is evaluated you don't need the `let` at all.

you can do `const rosterId` and then return it after the `forEach` call.

however if you want this function to be fully synchronous you need to do this:

```ts
await Promise.all(artifacts.map(async (a) => await createOrUpdateArtifact(rosterId, a)))
```

Reply:
I think this is what you mean? I've updated it for Artifacts for now to make sure I understand correctly